## **Data Cleaning in Pandas**

More often than not you will mostly have a dataset that you have close to little or no information about. In such cases your first step is to explore the data and get familiar with it. What are the columns ? How many observations do we have? What are the data types of each columns? Are there any duplicated values ? Do we have any missing values ? Any existing outliers? If we have user level data, how can we explore aggregate trends such as gender, race or geography. All of this questions can be answered by applying what we call **ETL(Extract Transform Load)**. ETL is the process by which you clean and transform your data into a suitable format. Applying ETL transforms raw datasets into alternative useful views. The primary choice of tool for use is the Python Package. Clean and well-structured data is crucial for accurate analysis , modelling and decision-making in data science and business intelligence. Let's go ahead and dive into the different techniques used to clean data.

### **Dealing with Missing Data**

Missing data can be problematic during the data science process because <font color = 'red'>NaN</font> values in the dataset limit our ability to do important things like:
- convert data types
- build models
- calculate summary statistics
- visualize data

### Detecting Missing Data

There two main ways missing data is often represented in datasets

**<font color = 'red'>NaN</font>**

By default, pandas represents null values with <font color = 'red'>NaN</font>, which is short form for **Not a Number**. Pandas provides many great ways for checking null values. These values are built right into the pandas DataFrame and Series objects.

### Detecting <font color = 'red'>NaN</font>

We will be using the titanic dataset. Go ahead and download the data from **[Here](https://github.com/atienosonia/PwaniTeknowGirls/blob/master/train.csv)**


In [ ]:
# import necessary libraries

import pandas as pd

# load data

titanic_data = pd.read_csv('/content/train.csv')

# to check whether a dataframe contains missing values
# we use the funtion df.isna() or df.isnull()
# let's go ahead and try this

# check for missing values

titanic_data.isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

Returns a matrix of boolean values where all cells containing <font color='red'>NaN</font> are converted to <font color='red'> True </font>and all cells containing valid data are converted to <font color='red'>False</font>

In [ ]:
# Let's proceed to sum the total data that is not valid

titanic_data.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Since <font color='red'> True </font> is equivalent to <font color='red'> 1 </font> and <font color='red'>False</font> is equivalent to <font color='red'>0</font> in Python, taking the <font color='red'>.sum()</font> of the DataFrame (or Series) will return the total number of NaN values in the dataset. Pandas even breaks this down by column.

### Placeholder Values


Often, datasets will contain missing values that are denoted by a value that seems valid to pandas. This is very common in real-world datasets. Usually, you find that the individuals entering the data are required to enter values that they don't actually have, so they enter an agreed-upon placeholder value.

The easiest way to deal with this is to familiarize yourself with the data dictionary that corresponds to your dataset--any placeholder values meant to denote a <font color='red'>NaN</font> value will be specified here.

However, you will encounter plenty of datasets in the real world that don't come with a data dictionary, or that fail to mention placeholders.

Follow the below strategies for detecting placeholder values in your data:

#### Numerical Data

Numerical columns will often represent missing values with a value that is nonsensical to the column in question. For instance, in healthcare data, missing values in a <font color ='red'>Weight</font> column may be using impossible values such as <font color ='red'>0</font> or <font color ='red'>9999</font> . These are valid to the computer, since they are real-numbered, but are obvious to anyone analyzing the data as placeholder values.

Such placeholder values are the most difficult to check for due to the fact that it requires getting decently familiar with the column in question to notice values that are technically valid but pragmatically impossible.

To detect these sorts of placeholder values, start by checking for outliers -- they are often encoded as very large numbers, or as 0 (when 0 isn't a realistic value for the column in question). **Any good placeholder value will be a value that couldn't show up in the real world.**

However, you may also find actual outliers (values that are not impossible, but occur rarely). Outliers, like missing or corrupt data, can adversely effect your machine learning models

#### Categorical Data

To detect placeholder values in categorical data, get the unique values in the column and see if there are any values that don't match up with your expectations.

Let's go ahead and try this with the Embarked column of the dataset.

In [ ]:
# get unique values in the Embarked column

titanic_data['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

*Note that for categorical columns, it is much more likely to have a data dictionary to work with, since it is common to have categorical values that aren't readily understandable without a data dictionary to help us figure out what each potential category means.*

#### Strategies for dealing with missing data

We have three options for dealing with missing data, you can either remove them, replace them or keep them.

#### **Remove**
The easiest way to deal with missing data is to drop the offending rows and/or columns. The disadvantage of this is that we lose data in the process. This is a valid strategy when dealing with large datasets -- however with smaller datasets throwing away data my be unacceptable. The two main strategies for dealing with missing values are to drop columns or to drop rows.

##### **Removing Columns**

Consider the ouput from the titanic dataset while checking the sum of missing values. The <font color='red'>Cabin</font> column contains 687 missing values. The entire dataset only contains about 900 entries/rows of data. In this case, it makes more sense to just remove the <font color='red'>Cabin</font> column from the entire dataset.

While this makes sense for the <font color='red'>Cabin</font> column, this is not a good idea for dealing with the null values contained within the <font color='red'>Age</font> column. Although the <font color='red'>Age</font> column contains 177 missing values, the vast majority of the items in this dataset contain perfectly good information for the age column. If we dropped this column, we would be throwing out all that information just to deal with a small subset of missing values in that column!

##### **Dropping Rows**

In th example above, dropping all rows that contain a null value may not be the smartest idea, dropping rows with missing values would mean losing 3/4 of our data ! Dropping rows make more sense when the proportion of rows with missing values is very small when compared to the size of the overall dataset.It's okay to just throw out the missing values as long as its not too many observations. There's no hard rule for exactly how many missing vales is the right amount to throw out, and will vary project by project. Think critically, and use your best judgement !

#### **Replace**

We can also deal with missing values by replacing them with a common value.The downside to this method is that it can introduce noise to our dataset

##### **Continous Data**

For continuous data, the best solution is to replace the missing values with the median value for that column. The median value is a good choice because it is least likely to influence the distribution of the overall dataset. If the dataset is symmetric, then the mean and the median will be the same value. If the dataset is not symmetric, then the mean is more likely to be skewed by outlier values, so the median is a better choice.

Pandas provides an easy way for us to replace null values. For instance, if the Fare column had missing values and we wanted to replace them with the median value, below is how we would replace it.

**<font color='red'>titanic_data['Fare'].fillna(titanic_data['Fare'].median())**

##### **Categorical Data**

With categorical data, this is harder, since we don't have summary statistics to lean on such as the median or the mean. In this case, if one categorical value is much more common than others, it is a valid strategy to replace missing values with this common value. However, make sure to examine your data first! If all the categorical values are equally common, picking one to replace all the missing values may do more harm than good by skewing the distribution and introducing some false signal into your dataset.

#### **Keep**

At times, the knowledge that a value is missing can itself be informative for us. If knowing that a value is missing tells us something then its often worth keeping the missing values using the following criteria:

##### **Categorical Data**

This one is the easiest--just treat missing values as its own category! This may require replacing missing values with a string to denote this, as your model will still likely throw errors if the actual <font color = 'red'>NaN</font> values are not replaced. In that case, just replace the <font color ='red'>NaN</font> values with the string <font color ='red'>'NaN'</font>, or another string that makes it obvious that this value is <font color = 'red'>'missing'</font>

##### **Numerical Data**

Often, missing values inside a continuously-valued column will cause all sorts of havoc in your models, so leaving the NaNs alone isn't usually an option here. Instead, consider using **Coarse Classification**, also referred to as **Binning**. This allows us to convert the entire column from a numerical column to a categorical column by binning our data into categories. For instance, we could deal with the missing values in the <font color ='red'>Age</font> column by creating a categorical column that separates each person into 10-year age ranges. Anybody between the ages of 0 and 10 would be a <font color = 'red'>1</font>, 11 to 20 would be a <font color = 'red'>2</font>, and so on.

Once we have binned the data in a new column, we can throw out the numerical version of the column, and just leave the missing values as one more valid category inside our new categorical column!

### **Check for Duplicates**

Duplicated elements occur when a string, integer or items in a list are repeated more than one time. We can check for duplicates using the function **df.duplicated()**

Let's try this using the titanic dataset

In [ ]:
# check for duplicates

titanic_data.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Length: 891, dtype: bool

In [ ]:
# preview the duplicated entries in a dataframe
# assign the variable duplicates to duplicated data

duplicates = titanic_data[titanic_data.duplicated()]

# print the length of duplicates in the dataset

print(len(duplicates))

0


This dataset seems to not have any duplicated entries which is great. It easens your data cleaning process. In the case where you get duplicated entries, you can drop them using **df.drop_duplicates(inplace = True)**. The inplace argument makes sure that the changes are made permanently to our dataframe.


### **Outliers**

An **outlier** is a data-item that deviates significantly from the rest of the (so called normal)data objects. There many ways of detecting outliers such as data visualization, using standard deviation, interquartile range, z-score and many others.


### **Pandas Groupby**

The **Pandas Grouby** method is used for grouping data according to categories. **Groupby** allows adopting a split-apply combine approach to a dataset. This approach is often used to slice and dice data in such a way that you can answer a specific question. Pandas objects can be split on any of their axes.

On a high level groupby allows to:

1. Split the data based on columns/conditions into groups.

2. Apply a function/transformation to all the groups and combine the results into an ouput

### **Using <font color = 'red'> .groupby()**</font>

Consider the titanic dataframe:


In [ ]:
# call the titanic dataset dataframe

titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

One of the common tasks you will want to do while performing Exploratory Data Analysis is spliting your dataset into subgroups and compare them to see if you can notice any trends. For instance, you many want to group the passengers together by Age or Gender. You can do this by using the <font color = 'red'> .groupby()</font> method built into the pandas DataFrames.

Let's go ahead and group passengers by gender

In [ ]:
# group the titanic dataset by gender

titanic_data.groupby('Sex')

Note that this alone will not display any results--Although you have split the dataset into groups, you don't have a meaningful way to display the information until you chain an **Aggregation Function** onto the groupby . This allows you to compute summary statistics

You can quickly use an aggregation function by chaining the call to the end of the <font color='red'>.groupby()</font> method.

In [ ]:
# group the titanic dataset by Sex

titanic_data.groupby('Sex').sum(numeric_only=True)

PassengerId  Survived  Pclass       Age  SibSp  Parch        Fare
Sex                                                                      
female       135343       233     678   7286.00    218    204  13966.6628
male         262043       109    1379  13919.17    248    136  14727.2865

#### **Aggregation Functions**

Some of the most common aggregate functions you many want to use are:

- <font color='red'>.min()</font>: returns the minimum value for each column by group
- <font color='red'>.max()</font>: returns the maximum value for each column by group
- <font color='red'>.mean()</font>: returns the average value for each column by group
- <font color='red'>.median()</font>: returns the median value for each column by group
- <font color='red'>.count()</font>: returns count of each column by group

#### **Multiple Groups**

You can also split data into multiple different levels of groups by passing in a list containing the name of every column you want to group-- for instance, by every combination of both <font color='red'>Sex</font> and <font color = 'red'>Pclass</font>

In [ ]:
# group the titanic dataset

titanic_data.groupby(['Sex', 'Pclass']).mean(numeric_only=True)

PassengerId  Survived        Age     SibSp     Parch  \
Sex    Pclass                                                         
female 1        469.212766  0.968085  34.611765  0.553191  0.457447   
       2        443.105263  0.921053  28.722973  0.486842  0.605263   
       3        399.729167  0.500000  21.750000  0.895833  0.798611   
male   1        455.729508  0.368852  41.281386  0.311475  0.278689   
       2        447.962963  0.157407  30.740707  0.342593  0.222222   
       3        455.515850  0.135447  26.507589  0.498559  0.224784   

                     Fare  
Sex    Pclass              
female 1       106.125798  
       2        21.970121  
       3        16.118810  
male   1        67.226127  
       2        19.741782  
       3        12.661633

#### **Selecting Information from Grouped Objects**

Since the resulting object returned is a DataFrame, you can also slice a selection of columns you're interested in from the DataFrame returned.

The example below demonstrates the syntax for returning the mean of the Survived class for every combination of <font color='red'>Sex</font> and <font color='red'>Pclass</font>

In [ ]:
# Calculates the average survival rate for each gender and passenger class combination.

titanic_data.groupby(['Sex', 'Pclass'])['Survived'].mean()

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64

### **Transforming Columns**

#### **Cleaning Column Names**

We will go ahead and use the MTA turnstile data which can be found **[Here](https://github.com/atienosonia/PwaniTeknowGirls/blob/master/turnstile_180901.txt).**This data is good when it comes to getting our hands dirty and doing some cleaning

In [ ]:
# import the necessary libraries

import pandas as pd

# load the data

turn_data = pd.read_csv('/content/turnstile_180901.txt', dtype=str)
turn_data.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  16:00:00   

      DESC     ENTRIES  \
0  REGULAR  0006736067   
1  REGULAR  0006736087   
2  REGULAR  0006736105   
3  REGULAR  0006736180   
4  REGULAR  0006736349   

  EXITS                                                                 
0  0002283184                                    ...                    
1  0002283188                                    ...                    
2  0002283229                                    ...                    
3  0002283314                                    ...                    
4  0002283384                                    ...

In [ ]:
# take a look at the columns in the data

turn_data.columns

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES',
       'EXITS                                                               '],
      dtype='object')

You notice that the EXITS column has a lot of annoying whitespace following. We can use a list comprehension to clean up all the columns

In [ ]:
# use a list comprehension to clean up the column names

[col.strip() for col in turn_data.columns]

['C/A',
 'UNIT',
 'SCP',
 'STATION',
 'LINENAME',
 'DIVISION',
 'DATE',
 'TIME',
 'DESC',
 'ENTRIES',
 'EXITS']

We need to reassign turn_data.columns for the changes to be stored in memory


In [ ]:
# reassign the list comprehension to turn_data.columns

turn_data.columns = [col.strip() for col in turn_data.columns]
turn_data.columns

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES', 'EXITS'],
      dtype='object')

#### **Cleaning Column Names**

You can also rename column names using dictionaries. Let's say we want to rename C/A to CONTROL_AREA(the data dictionary indicates that this is what it represents) and UNIT to REMOTE_UNIT. Access the data dictionary from **[Here](https://github.com/atienosonia/PwaniTeknowGirls/blob/master/turnstile_180901.txt).**

In [ ]:
# rename the C/A column

turn_data.rename(columns = {'C/A': 'CONTROL_AREA',
                            'UNIT' : 'REMOTE_UNIT'})

CONTROL_AREA REMOTE_UNIT       SCP   STATION LINENAME DIVISION  \
0              A002        R051  02-00-00     59 ST  NQR456W      BMT   
1              A002        R051  02-00-00     59 ST  NQR456W      BMT   
2              A002        R051  02-00-00     59 ST  NQR456W      BMT   
3              A002        R051  02-00-00     59 ST  NQR456W      BMT   
4              A002        R051  02-00-00     59 ST  NQR456W      BMT   
...             ...         ...       ...       ...      ...      ...   
103263         N554        R423  01-04-00  AVENUE N        F      IND   
103264         N554        R423  01-04-00  AVENUE N        F      IND   
103265         N554        R423  01-04-00  AVENUE N        F      IND   
103266         N554        R423  01-04-00  AVENUE N        F      IND   
103267         N554        R423  01-04-00  AVENUE N        F      IND   

              DATE      TIME     DESC     ENTRIES  \
0       08/25/2018  00:00:00  REGULAR  0006736067   
1       08/25/2018  04:00:00  REGULAR  0006736087   
2       08/25/2018  08:00:00  REGULAR  0006736105   
3       08/25/2018  12:00:00  REGULAR  0006736180   
4       08/25/2018  16:00:00  REGULAR  0006736349   
...            ...       ...      ...         ...   
103263  08/26/2018  09:00:00  REGULAR  0134217900   
103264  08/26/2018  13:00:00  REGULAR  0134217900   
103265  08/26/2018  17:00:00  REGULAR  0134217900   
103266  08/26/2018  21:00:00  REGULAR  0134217900   
103267  08/27/2018  01:00:00  REGULAR  0134217900   

                                                    EXITS  
0       0002283184                                    ...  
1       0002283188                                    ...  
2       0002283229                                    ...  
3       0002283314                                    ...  
4       0002283384                                    ...  
...                                                   ...  
103263  0000000000                                    ...  
103264  0000000000                                    ...  
103265  0000000000                                    ...  
103266  0000000000                                    ...  
103267            0000000000                               

[103268 rows x 11 columns]

Again, note that the DataFrame was not fully transformed by doing this. If we look at it now, C/A and UNIT are still there:

In [ ]:
# preview the dataframe

turn_data.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018  16:00:00   

      DESC     ENTRIES                                              EXITS  
0  REGULAR  0006736067  0002283184                                    ...  
1  REGULAR  0006736087  0002283188                                    ...  
2  REGULAR  0006736105  0002283229                                    ...  
3  REGULAR  0006736180  0002283314                                    ...  
4  REGULAR  0006736349  0002283384                                    ...

If we want to make the changes stick, one way to do this is to use the argument <font color='red'>'inplace = True'</font> to the rename function

In [ ]:
# rename the columns and make the changes permanent

turn_data.rename(columns = {'C/A':'CONTROL_AREA',
                  'UNIT': 'REMOTE_UNIT'}, inplace = True)

Proceed to check if this has been implemented

In [ ]:
# call the dataframe

turn_data.head()

CONTROL_AREA REMOTE_UNIT       SCP STATION LINENAME DIVISION        DATE  \
0         A002        R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018   
1         A002        R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018   
2         A002        R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018   
3         A002        R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018   
4         A002        R051  02-00-00   59 ST  NQR456W      BMT  08/25/2018   

       TIME     DESC     ENTRIES  \
0  00:00:00  REGULAR  0006736067   
1  04:00:00  REGULAR  0006736087   
2  08:00:00  REGULAR  0006736105   
3  12:00:00  REGULAR  0006736180   
4  16:00:00  REGULAR  0006736349   

                                               EXITS  
0  0002283184                                    ...  
1  0002283188                                    ...  
2  0002283229                                    ...  
3  0002283314                                    ...  
4  0002283384                                    ...

**Note that this behavior (not changing the contents of the dataframe unless you use inplace=True or reassign the variable) is not a mistake or oversight in pandas. It is a useful feature that lets you preview the outcome of an operation before permanently applying it! This is especially important if you are dropping data or transforming it in a way that is not reversible.**

#### **Dropping Column Names**

Let's say we have determined that the DESC column doesn't matter. We can test out dropping it like this.

In [ ]:
# drop the column DESC

turn_data.drop('DESC', axis = 1)

CONTROL_AREA REMOTE_UNIT       SCP   STATION LINENAME DIVISION  \
0              A002        R051  02-00-00     59 ST  NQR456W      BMT   
1              A002        R051  02-00-00     59 ST  NQR456W      BMT   
2              A002        R051  02-00-00     59 ST  NQR456W      BMT   
3              A002        R051  02-00-00     59 ST  NQR456W      BMT   
4              A002        R051  02-00-00     59 ST  NQR456W      BMT   
...             ...         ...       ...       ...      ...      ...   
103263         N554        R423  01-04-00  AVENUE N        F      IND   
103264         N554        R423  01-04-00  AVENUE N        F      IND   
103265         N554        R423  01-04-00  AVENUE N        F      IND   
103266         N554        R423  01-04-00  AVENUE N        F      IND   
103267         N554        R423  01-04-00  AVENUE N        F      IND   

              DATE      TIME     ENTRIES  \
0       08/25/2018  00:00:00  0006736067   
1       08/25/2018  04:00:00  0006736087   
2       08/25/2018  08:00:00  0006736105   
3       08/25/2018  12:00:00  0006736180   
4       08/25/2018  16:00:00  0006736349   
...            ...       ...         ...   
103263  08/26/2018  09:00:00  0134217900   
103264  08/26/2018  13:00:00  0134217900   
103265  08/26/2018  17:00:00  0134217900   
103266  08/26/2018  21:00:00  0134217900   
103267  08/27/2018  01:00:00  0134217900   

                                                    EXITS  
0       0002283184                                    ...  
1       0002283188                                    ...  
2       0002283229                                    ...  
3       0002283314                                    ...  
4       0002283384                                    ...  
...                                                   ...  
103263  0000000000                                    ...  
103264  0000000000                                    ...  
103265  0000000000                                    ...  
103266  0000000000                                    ...  
103267            0000000000                               

[103268 rows x 10 columns]

Notice the axis=1 argument ? By default, df.drop() tries to drop rows (axis=0). If you try dropping a column and you forget the axis argument then you will get an error

#### **Changing Column Types**

Another common data munging technique can be reformatting column types. We first viewed the column types using df.info(). Let's go ahead and repeat that here


In [ ]:
# check the info of the dataframe

turn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103268 entries, 0 to 103267
Data columns (total 11 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   CONTROL_AREA  103268 non-null  object
 1   REMOTE_UNIT   103268 non-null  object
 2   SCP           103268 non-null  object
 3   STATION       103268 non-null  object
 4   LINENAME      103268 non-null  object
 5   DIVISION      103268 non-null  object
 6   DATE          103268 non-null  object
 7   TIME          103268 non-null  object
 8   DESC          103268 non-null  object
 9   ENTRIES       103268 non-null  object
 10  EXITS         103268 non-null  object
dtypes: object(11)
memory usage: 8.7+ MB


We can also check the data type of an individual column rather than listing all of them.

In [ ]:
# check the data type of the ENTRIES column

print(turn_data['ENTRIES'].dtype)

object


In this case we specified **dtype=str** when we opened the file, telling pandas to treat all of the columns as strings initially. So currently every column except for **On_N_Line** is dtype **object**

A common transformation needed is converting the data type stored as text (dtype = object) into an integer or float representations.

Let's look more closely at **ENTRIES**

In [ ]:
# preview the first 5 rows of the column ENTRIES
# use label indexing

turn_data.loc[:5, 'ENTRIES']

0    0006736067
1    0006736087
2    0006736105
3    0006736180
4    0006736349
5    0006736562
Name: ENTRIES, dtype: object

These seem like integers to me, let's try converting the data type.
We will go ahead and convert the whole column

In [ ]:
# change the data type of the ENTRIES column from object to an integer

turn_data['ENTRIES'] = turn_data['ENTRIES'].astype(int)

In [ ]:
# check the dataframe info

turn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103268 entries, 0 to 103267
Data columns (total 11 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   CONTROL_AREA  103268 non-null  object
 1   REMOTE_UNIT   103268 non-null  object
 2   SCP           103268 non-null  object
 3   STATION       103268 non-null  object
 4   LINENAME      103268 non-null  object
 5   DIVISION      103268 non-null  object
 6   DATE          103268 non-null  object
 7   TIME          103268 non-null  object
 8   DESC          103268 non-null  object
 9   ENTRIES       103268 non-null  int64 
 10  EXITS         103268 non-null  object
dtypes: int64(1), object(10)
memory usage: 8.7+ MB


#### **Setting a New Index**
It can also be helpful to set one of the columns as the index of the DataFrame, such as when graphing

In [ ]:
# set the index to DATE column

turn_data = turn_data.set_index('DATE')
turn_data.head()

CONTROL_AREA REMOTE_UNIT       SCP STATION LINENAME DIVISION  \
DATE                                                                      
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   

                TIME     DESC  ENTRIES  \
DATE                                     
08/25/2018  00:00:00  REGULAR  6736067   
08/25/2018  04:00:00  REGULAR  6736087   
08/25/2018  08:00:00  REGULAR  6736105   
08/25/2018  12:00:00  REGULAR  6736180   
08/25/2018  16:00:00  REGULAR  6736349   

                                                        EXITS  
DATE                                                           
08/25/2018  0002283184                                    ...  
08/25/2018  0002283188                                    ...  
08/25/2018  0002283229                                    ...  
08/25/2018  0002283314                                    ...  
08/25/2018  0002283384                                    ...

#### **Having the Columns in Lower Case**

You may also want to have your columns in lower case for consistency and readability. Also to avoid case sensitive issues. Below is how you can go ahead and do that:

In [ ]:
# use a list comprehension to have the column names in all lower case

[col.lower() for col in turn_data.columns]

# assign column names to the list comprehension

turn_data.columns = [col.lower() for col in turn_data.columns]

# preview the data

turn_data.head()

control_area remote_unit       scp station linename division  \
DATE                                                                      
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   
08/25/2018         A002        R051  02-00-00   59 ST  NQR456W      BMT   

                time     desc  entries  \
DATE                                     
08/25/2018  00:00:00  REGULAR  6736067   
08/25/2018  04:00:00  REGULAR  6736087   
08/25/2018  08:00:00  REGULAR  6736105   
08/25/2018  12:00:00  REGULAR  6736180   
08/25/2018  16:00:00  REGULAR  6736349   

                                                        exits  
DATE                                                           
08/25/2018  0002283184                                    ...  
08/25/2018  0002283188                                    ...  
08/25/2018  0002283229                                    ...  
08/25/2018  0002283314                                    ...  
08/25/2018  0002283384                                    ...